### ISSUE TITLE
How to preprocess the data without using dataset? And how to match embeddings?

### ISSUE THREAD

**Issue by hcl14** [MadryLab/robust_representations#2]
Hi, thanks for the great work! 

I wanted to test the model on some of my custom images, and tried to understand what preprocessing steps were applied. By examining outputs of the dataset function, I found that all of them have min=0 and max=1. Also I found `TEST_TRANSFORMS_224` list and that ImageNet mean and std are present in the dataset function, but I wonder if they are used. Does the following way to process input image reflect all the transformations I need for ImageNet model?

```
from PIL import Image
import torch as ch
from torchvision import transforms

TEST_TRANSFORMS_224 = transforms.Compose([
        transforms.ToPILImage(),
        transforms.Resize((224,224)),
        transforms.ToTensor(), # This converts the image to range [0,1]
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
    ])

def normalize(vector):
    min_v = ch.min(vector)
    range_v = ch.max(vector) - min_v
    if range_v > 0:
        normalised = (vector - min_v) / range_v
    else:
        normalised = ch.zeros(vector.size())
    return normalised


def get_patch_embedding(np_img):
   im_1 =  TEST_TRANSFORMS_224(ch.from_numpy(np_img)).unsqueeze(0)
   im_1 = normalize(im_1.cuda())
   (_, rep_1), _  = model(im_1, with_latent=True)
   return rep_1.detach().cpu().numpy()

im_1 = np.array(Image.open('9_256_384.png').convert('RGB'))

emb = get_patch_embedding(im_1)
```

Another question is how correct would be to compare embeddings `emb1`, `emb2` with l2, cosine, etc?

**Comment by lengstrom** [comment]
Sounds cool! 

Currently your method has at least two problems.

- Our model requires the original image in [0,1] rgb format. We normalize internally within the model. Your method would not work; the normalize function is not an inverse of `transforms.Normalize`; you should take out `transforms.Normalize` and `normalize` for a correct implementation. 
- You need to scale the largest side to 224, then square center crop to (224,224) instead of stretching the image.

There may be more problems here, I can't certify that this is correct - why not just use our code for this preprocessing?

> Another question is how correct would be to compare embeddings `emb1`, `emb2` with l2, cosine, etc?

Whether or not each of these is correct depends on what your goal is.
